AX: Implement support for serving attributed strings off the main-thread for ranges that span multiple elements

twilco opened the pull request:
#### 0d448f9e4bf09635515def81acc656df0fed5d31
<pre>
AX: Implement support for serving attributed strings off the main-thread for ranges that span multiple elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=284193">https://bugs.webkit.org/show_bug.cgi?id=284193</a>
<a href="https://rdar.apple.com/problem/141067345">rdar://problem/141067345</a>

Reviewed by Chris Fleizach.

This commit implements support for serving attributed strings off the main-thread for ranges that span multiple elements.
Similar to AXTextMarkerRange::toString(), we traverse from the object starting the range to the end, building an attributed
string as we go.

This patch also fixes a bug in the off-main-thread text marker implementation where we did not emit two newlines for
paragraph elements, and a tab for table cells as mandated by the spec (<a href="https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute).">https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute).</a>

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-spanning-multiple-elements-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-spanning-multiple-elements.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::emitsNewlineAfter const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::Accessibility::findObjectWithRuns):
(WebCore::findObjectWithRuns): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::Accessibility::findObjectWithRuns):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::emitTextAfter const):
(WebCore::AccessibilityNodeObject::shouldEmitNewlinesBeforeAndAfterNode const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: <a href="https://commits.webkit.org/287745@main">https://commits.webkit.org/287745@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edd645fe30992f1a1f37829bac1fde6618b419f9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59091 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84606 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82198 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68152 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62612 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20438 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83157 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52694 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42915 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50023 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27106 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29521 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71147 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27613 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86038 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7309 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5178 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70892 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68790 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70125 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17579 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14123 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13077 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12790 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7112 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10632 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8917 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
```diff
@@ -87,8 +90,34 @@
+        [result appendAttributedString:current->createAttributedString(current->textRuns()->toString(), AXCoreObject::SpellCheck::No).get()];
+        current = findObjectWithRuns(*current, AXDirection::Next, std::nullopt, emitNewlineOnExit);
+    }
+    [result appendAttributedString:end.isolatedObject()->createAttributedString(end.runs()->substring(0, end.offset()), AXCoreObject::SpellCheck::No).get()];
```
Does the result of createAttributedString() need to be autoreleased?

twilco replied:
For these temporary strings, I think so — good call! I added that in the latest commit. Let me know if you think I'm missing any else.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1101,12 +1103,36 @@ AccessibilityButtonState AccessibilityNodeObject::checkboxOrRadioValue() const
 
 #if ENABLE(AX_THREAD_TEXT_APIS)
-bool AccessibilityNodeObject::shouldEmitNewlinesBeforeAndAfterNode() const
+TextEmissionBehavior AccessibilityNodeObject::emitTextAfter() const
```
Nit: the name of this reads like a bool to me, but returns a `TextEmissionBehavior`. What do you think about calling this `textEmissionBehavior()`, or `emitTextAfterBehavior()`?

twilco replied:
Good point, `emitTextAfterBehavior` sounds better, will go with that! I anticipate we'll eventually need `emitTextBeforeBehavior`, hence why `textEmissionBehavior` may not work as well.

twilco committed:
AX: Implement support for serving attributed strings off the main-thread for ranges that span multiple elements
https://bugs.webkit.org/show_bug.cgi?id=284193
rdar://problem/141067345

Reviewed by Chris Fleizach.

This commit implements support for serving attributed strings off the main-thread for ranges that span multiple elements.
Similar to AXTextMarkerRange::toString(), we traverse from the object starting the range to the end, building an attributed
string as we go.

This patch also fixes a bug in the off-main-thread text marker implementation where we did not emit two newlines for
paragraph elements, and a tab for table cells as mandated by the spec (https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute).

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-spanning-multiple-elements-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-spanning-multiple-elements.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::emitsNewlineAfter const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::toString const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::Accessibility::findObjectWithRuns):
(WebCore::findObjectWithRuns): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::Accessibility::findObjectWithRuns):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::emitTextAfter const):
(WebCore::AccessibilityNodeObject::shouldEmitNewlinesBeforeAndAfterNode const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: https://commits.webkit.org/287745@main
